src/main/jbake/content/blog/2017/06/23/opentracing-appmetrics.adoc

```diff
@@ -0,0 +1,270 @@
+service.
+
+
+== Deploying on Kubernetes
+
+The steps to setup an environment on Kubernetes is discussed in the
+https://github.com/objectiser/opentracing-prometheus-example[example codebase]. A summary of the steps is:
+
+* Start https://kubernetes.io/docs/getting-started-guides/minikube[minikube]
+
+* Deploy Prometheus - using the https://github.com/coreos/prometheus-operator[Prometheus Operation] project to
+capture metrics from the services
+
+* Deploy https://github.com/uber/jaeger[Jaeger] - an OpenTracing compatible tracing system
+
+* For this article, we also https://github.com/kubernetes/charts/tree/master/stable/grafana[deployed Grafana]
+to display the metrics, although the Prometheus dashboard could be used.
+
+Once they are all running, then the simple example with the two services can be deployed. At this stage
+the Kubernetes dashboard would look like this:
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-06-23-kubernetes-dashboard.png[caption="Figure 1: ", title="Kubernetes dashboard"]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-06-23-kubernetes-dashboard.png[caption="Figure 1: ", title="Kubernetes dashboard"]
+endif::[]
+
+The example code includes a script that loops randomly invoking the three REST endpoints provided by
+_service1_. Once some example requests have been created, you can view the tracing dashboard:
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-06-23-traces.png[caption="Figure 2: ", title="Jaeger tracing dashboard"]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-06-23-traces.png[caption="Figure 2: ", title="Jaeger tracing dashboard"]
+endif::[]
+
+Then you can select a specific trace instance and see further details:
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-06-23-trace.png[caption="Figure 3: ", title="Jaeger trace instance view"]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-06-23-trace.png[caption="Figure 3: ", title="Jaeger trace instance view"]
+endif::[]
+
+This shows that the trace instance has three spans, the first representing the receipt of the `/buy` request on
+_service1_, the second where _service1_ is invoking _service2_, and finally the _service2_ receiving the
+`/hello` request. In this particular trace instance, the _service2_ invocation has reported an error, indicated
+by the `error=true` tag.
+
+Now we will look at the Grafana dashboard to see what metrics have been reported from the OpenTracing
+instrumentation within the two services:
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-06-23-grafana-dashboard.png[caption="Figure 4: ", title="Grafana dashboard"]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-06-23-grafana-dashboard.png[caption="Figure 4: ", title="Grafana dashboard"]
+endif::[]
+
+This dashboard includes three graphs, the first showing the span count, the second showing the average duration
```

Comment:
If this is a particular span type, like "sell" or "buy",  then "business operation" might be an alternative. Something like: 

> the first showing the span count for spans created by our `sell()` method, and we can use it to track how many times this business operation has been executed. The second, ...